### ISSUE TITLE
Questions about integrating with Mathy environments

### ISSUE THREAD

**Issue by justindujardin** [Guillemdb/fragile#13]
Thanks again for sharing your awesome research! I spent some time over the weekend integrating the Fragile library with Mathy. I was able to get the code to run and step through environments, but I have run into some issues that I'm not sure how best to solve.

*the draft PR for mathy is available [here](https://github.com/justindujardin/mathy/pull/30) if you're interested, and [here](https://github.com/justindujardin/mathy/blob/feature/fragile/mathy_fragile.py)'s the script that runs the demo.*

### Masked actions

Mathy uses a masked action-space to reduce the number of actions it has to choose from because it's space is sparsely populated with valid actions. Does Fragile support (or could it?) masked action spaces?

Specifically, the action-space for Mathy is `(num_rules * num_nodes)` where num_rules is the number of available actions, and num_nodes is the number of nodes in the expression tree. As you can guess, for large expressions (some have > 100 nodes) the action space can grow to 500 or more possible actions. On top of that, the number of valid actions is usually very small. Consider that you can't validly apply the distributive property to most nodes in a math tree.

If fragile doesn't work with masked action spaces, how would you recommend dealing with an action space like Mathy's that has a variable number of actions, where only a small portion of them are valid in any given state?

### Short Episodes

For Mathy the goal is for the agent to select the *ideal* action at every timestep so that the user can follow along with the agent's work step-by-step. In order to use the step history as instructions, mathy tries to optimize for as few steps as possible in episodes.

For example, it's considered an error for an agent to perform the associative swap a bunch of times on unimportant nodes before combining constants to reach a goal state. Just solving a problem isn't really good enough, the agents need to solve them with as few moves as possible. Have you experimented with this kind of setup at all?

---

Thanks for any thoughts you have, I'll consider these things and try to get better results next weekend.

**Comment by justindujardin** [comment]
Going to close this so it doesn't clutter the repo issues. It's mostly a usage question with no real action needed for resolution.

**Comment by Guillemdb** [comment]
Hi Justin, 

Seeing your code just made me realize that I should add the condition of finishing after finding a maximum reward to the core interface. You shouldn't be wasting time on that, sorry.

I will be adding tutorials on how to integrate new environments and models after the first release is ready, but I write a short answer here.

## Masked actions

Given the sparsity and the characteristics of your state space, the most efficient way of implementing your masked action space is writing a Model that:
- reads the environment state (it should contain all the necessary information to do the masking)
- Calculates the masking for a given state and samples an action
- returns the sampled action as an integer (or some small data structure that the environment can decode). This avoids the computational overhead of copying big chunks of data when passing the states around.

I wouldn't mind adding support for masked spaces, but I have never used them before. I am not sure of what the best interface would be. Can you point me to the code where you define them so I can get a better idea of how they work?

## Adapting FAI to your specific state space

I advise you that after you make the environment and model work with the `fragile` interface, you don't use it yet.

 In your case the `Swarm` algorithm shouldn't give you a big improvement over MCTS (both a Swarm and MCTS will probly get bad results). State spaces that are very wide and shallow are the worst possible ones for a `Swarm`. The current algorithm pushes the walkers forward in time, while in your case they should be able to go back in time.

But don't worry. I want to release a new Go-explore like algorithm that should be suitable for your problem. If you manage to get the `Environment` and  `Model` to work with `mathy` you will be able to use the new algorithm once it's released.

In the meantime, if you want to improve the performance of the search algorithm I suggest that you try to define a suitable similarity function that measures "how different" two nodes in your search tree are. 

Another thing that can give you a huge boost, is defining boundary conditions on your search space. Quickly finding out and discarding useless regions of the state space will improve the convergence a lot. 

I am planning on making tutorials about this after the first release, but if you wanna try to do it yourself I can help you implement them.

**Comment by justindujardin** [comment]
Thanks for the ideas, I will think about them thoroughly! A few initial thoughts:

> Given the sparsity and the characteristics of your state space, the most efficient way of implementing your masked action space is writing a Model

I found the Model class in Fragile, and I think I can see how to do masked spaces using it. I'll try it out this weekend!

> I wouldn't mind adding support for masked spaces, but I have never used them before. I am not sure of what the best interface would be.

The masking in Mathy isn't as tidy as I would like, and it spans a few pieces of code. The basic setup is that the env provides a binary action mask as part of each observation, and that is used to mask out invalid actions during selection. 

_I originally got the idea from [AlphaZero](https://arxiv.org/pdf/1712.01815.pdf), but it's probably worth noting that the new variant [MuZero](https://arxiv.org/pdf/1911.08265.pdf) does away with masking except at the first timestep in a search._

> Can you point me to the code where you define them so I can get a better idea of how they work?

For the bits of code, first mathy [generates the action mask](https://github.com/justindujardin/mathy/blob/master/libraries/mathy_python/mathy/env.py#L446-L469) to include in an [observation](https://github.com/justindujardin/mathy/blob/478fad505195aa3ecdeebb84368b6b1a681744c1/libraries/mathy_python/mathy/state.py#L239-L264). Then the ML model [uses the mask](https://github.com/justindujardin/mathy/blob/master/libraries/mathy_python/mathy/agents/policy_value_model.py#L148-L166) to produce logits that will only select from valid indices. Finally, an [action is selected](https://github.com/justindujardin/mathy/blob/master/libraries/mathy_python/mathy/agents/policy_value_model.py#L168-L176) by applying softmax to the policy logits and selecting from the probs.

Also there's a [OpenAI gym wrapper that does masking](https://github.com/justindujardin/mathy/blob/master/libraries/mathy_python/mathy/envs/gym/masked_discrete.py) for an action space. I imagine something like this is what I'll need to do with your Model class.

> both a Swarm and MCTS will probly get bad results

Interestingly, the MCTS agent does quite well at solving complex tasks, but it uses masking and a neural-network trained prior to do its searching, which make it feel slow compared to the A3C agent. You can see the "zero" agent search/masking [here](https://github.com/justindujardin/mathy/blob/master/libraries/mathy_python/mathy/agents/mcts.py#L65-L92)

> In the meantime, if you want to improve the performance of the search algorithm I suggest that you try to define a suitable similarity function that measures "how different" two nodes in your search tree are.

I suspected that the distance function was important. I will think about the best way to compare states and make adjustments as needed. I am a little unsure about how I should represent the state of each node for comparison. Say I have two paths to the same end state:

```
a = [
   "4x + 2 + 3y", 
   "4x + 3y + 2", 
   "(4 + 3) * x + 2",
   "7x + 2"
]
b = [
   "4x + 2 + 3y", 
   "4x + 3y + 2", 
   "4x + 2 + 3y", 
   "4x + 3y + 2", 
   "4x + 2 + 3y", 
   "4x + 3y + 2", 
   "4x + 2 + 3y", 
   "4x + 3y + 2", 
   "(4 + 3) * x + 2",
   "7x + 2"
]
```

When comparing the distance between `a` and `b` end states, would it be right to include all the history so that even though they're in the same final state, the distance is large because of the extra steps that `b` took to get there? Or regardless of the path, they're the same final state, so their distance should be 0.0? _I suspect that because Mathy envs [give rewards based on how quickly you complete problems](https://github.com/justindujardin/mathy/blob/master/libraries/mathy_python/mathy/env.py#L152-L165), that the comparison should be made given the state and perhaps the current timestep. I'll think about this more._

> Another thing that can give you a huge boost, is defining boundary conditions on your search space. Quickly finding out and discarding useless regions of the state space will improve the convergence a lot.

I'm definitely interested in this, how do you actually do that in Fragile? Can you point to an example in the code?

**Comment by justindujardin** [comment]
I managed to get fragile to solve Mathy problems, though it's solutions probably aren't as nice as the improved Go-explore like algorithm you have in mind. 

In fact, the fragile swarm is even able to find solutions to the "hard" poly problems that mathy generates, which are difficult for a trained A3C agent to solve.

I had a change a few things to make it work:
 - treat invalid actions as no-ops in the environment and return 0.0 reward
 - reduce the GaussianDt to min/max of 1 so the walkers don't repeat actions
 - remove negative reward penalties and replace them with very small positive value rewards

I ran into two problems that I wanted to ask you about:

First, it appears that terminal rewards and timesteps are not included in the "best_reward" or tree paths. In Mathy environments whenever you solve a problem you get a large reward for the final timestep. I increased the end reward to a magnitude that it can be used for early-stopping, but the max-reward is never triggered because the terminal values are ignored. I tried to debug this, but all I was able to verify is that the walkers get the terminal rewards added to their states, but they don't get reflected in the stats or output tree paths. You can reproduce this error state easily by [running my script with any problem](https://github.com/justindujardin/mathy/blob/feature/fragile/mathy_fragile.py). The final reward is always left off, despite it being a 10.0 value, and the final "completed" state is missing from the output at the end.

The other problem is that I tried integrating the dataviz module that you added, and it doesn't seem to like running outside of a Jupyter notebook. The holowviews/bokeh documentation was not very good on how to run them outside a notebook, so I wonder if you have plans to support standalone visualization?

I haven't tried out masking the action space yet because ignoring invalid actions seems to have done well enough for the moment. For the poly/hard problems I have to turn the walkers up to 768 for good results, so I'll give the masking a go to see if I can reduce that number.

Thanks for your help so far!

**Comment by Guillemdb** [comment]
Sorry, I have been quite busy these past days. I will take time to answer everything this evening.

**Comment by justindujardin** [comment]
I totally understand being busy, don't rush to answer my questions. Whenever you have some spare time is fine by me.

**Comment by Guillemdb** [comment]
You are probably the only person using this library, so I will try to answer you ASAP :)

> For the bits of code, first mathy generates the action mask to include in an observation. Then the ML model uses the mask to produce logits that will only select from valid indices. Finally, an action is selected by applying softmax to the policy logits and selecting from the probs.
> 
> Also there's a OpenAI gym wrapper that does masking for an action space. I imagine something like this is what I'll need to do with your Model class.

Yest, looks that masking and calculating the action should be done inside the model. After taht store the index in the model_states sending the index of the environment.  Given that the non-masked vector of all possible actions doesn't look too big, you could also include it inside the model_states  if the Environment needs it (but it's a bit of additional overhead, and I have not tried using sparse arrays).

> When comparing the distance between a and b end states, would it be right to include all the history so that even though they're in the same final state, the distance is large because of the extra steps that b took to get there? Or regardless of the path, they're the same final state, so their distance should be 0.0? I suspect that because Mathy envs give rewards based on how quickly you complete problems, that the comparison should be made given the state and perhaps the current timestep. I'll think about this more.

Given that the algorithms are derived from the statistical mechanics, framework we consider each state markovian, but in practice the distance is only a function that measures "how different two states are".  The shortest answer is: "**Try both options and choose the one it works better**". However, in your case I would try this things and see what works better:
-  Not including the path information, but only the current nodes being compared. (You already take care of that in the reward)
-  An additional coordinate in your state vector equal to the path length.
-  Incorporating the information regarding path length in your reward (You can combine several terms to build it as long as you return a single scalar, and soon I will add classes for multi-objective problems)

> I suspect that because Mathy envs give rewards based on how quickly you complete problems, 

If you can only get a "true reward" after finishing a problem, I have an algorithm that uses a Swarm + and Actor-Critic that works extremely well (at least in neural networks architecture selection). Unfortunately I need to add a bunch of features before I can implement it. I am pretty sure that would be the optimal algorithm for your problem, but in the meantime I will release other algorithms that will work better than a regular Swarm.

> I'm definitely interested in this, how do you actually do that in Fragile? Can you point to an example in the code?

> treat invalid actions as no-ops in the environment and return 0.0 reward

The boundary condition is controlled by the `end` values of the environment. If a given state lies outside your defined boundaries make the end value `True` and the swarm will avoid it. This means that besides adding a 0.0 reward to you invalid actions, you assign them a `end = True` the walkers will avoid invalid states.

For example, the end flag of Atari games are treated as a custom boundary condition, but that can generalize to any kind of environment. If you need to differentiate between `end` as a death condition, or `end` as in the environment successfully finished, use the same trick I used in the  [AtariWalkers](https://github.com/FragileTech/fragile/blob/09853b927f35ad25065df398f4eaaf777cefa392/fragile/atari/walkers.py#L30) and treat the successfully finished states as `game_ends`. The [AtariEnvironement](https://github.com/FragileTech/fragile/blob/09853b927f35ad25065df398f4eaaf777cefa392/fragile/atari/env.py#L62) uses both, so you have an example of how to do it there. I can help you if you need it.

> reduce the GaussianDt to min/max of 1 so the walkers don't repeat actions

The time step is always one by deffault. You can skip using a `GaussianDt` altogether. I will modify the examples to make that clearer.

> remove negative reward penalties and replace them with very small positive value rewards

The `Swarm` is resilient to the sign of the reward, it only compares the relative scales of the rewards that two states have. Feel free to add a huge reward to good states and negative rewards to bad states, specially if you mask out the invalid moves with a death condition. (Otherwise an invalid state will be better than a one with negative reward.)

> First, it appears that terminal rewards and timesteps are not included in the "best_reward" or tree paths.

You are right. Now I am only storing the states and the actions because I was just caring of getting the pruning algorithm right and linear time performance. Once I am done with the test coverage I will add a data generation module, with trees to store abitrary parameters automatically, and generators to easily extract the saved data. It should be ready next week.

> The final reward is always left off, despite it being a 10.0 value, and the final "completed" state is missing from the output at the end.

This is probably because you are marking the final states with the `end` flag, when this flag is supposed to be for marking boundary conditions. Could you please suggest me a better name for `end` (arbitrary boundary condition flag), and a name for the value that marks the end of an environment run (now begin `game_ends` ? I will also make some refactoring and a tutorial to make it easier to understand. I have been using this for so long that I don't realize when my naming conventions are off, sorry.

> The other problem is that I tried integrating the dataviz module that you added, and it doesn't seem to like running outside of a Jupyter notebook. The holowviews/bokeh documentation was not very good on how to run them outside a notebook, so I wonder if you have plans to support standalone visualization?

The can also be visualized in a bokeh server (totally a PITA though). I haven't thought about this until you mentioned it :sweat_smile: . Plotting interactive visualizations with data streaing outside the notebook is impossible, but I could write a wrapper saves the plots as `.png` files in a target directory. Would that work for you?

**Comment by justindujardin** [comment]
Thanks for the quick response!

> This is probably because you are marking the final states with the end flag, when this flag is supposed to be for marking boundary conditions. Could you please suggest me a better name for end (arbitrary boundary condition flag), and a name for the value that marks the end of an environment run (now begin game_ends ? I will also make some refactoring and a tutorial to make it easier to understand. I have been using this for so long that I don't realize when my naming conventions are off, sorry.

Ahh, I see! What about `oob=True` (out-of-bounds) for boundary conditions, and `terminal=True` for game-ended states? OOB is something you'd see frequently in games, and Terminal is used in a bunch of the existing RL literature to mean an ended state.

> Yes, looks that masking and calculating the action should be done inside the model.

I got the masking to work by making the env observations a probability distribution over the valid action mask, then I choose an action at random with those probabilities. Since the algorithm doesn't need to "learn" anything for reuse, the probability distribution over actions as an input works quite well 😅 The swarm does much better now, only requiring about 1/3 the number of walkers it did previously to solve the hard polynomial problems.

> If you can only get a "true reward" after finishing a problem, I have an algorithm that uses a Swarm + and Actor-Critic that works extremely well (at least in neural networks architecture selection). Unfortunately I need to add a bunch of features before I can implement it. I am pretty sure that would be the optimal algorithm for your problem, but in the meantime I will release other algorithms that will work better than a regular Swarm.

I'm ultimately interested in mixing the swarm with a learned model, so I agree that this may be the optimal algorithm. When you get around to that, I'm happy to help in any way that I can.

Next, I'll try adjusting the end/game_ends and see if I can get the final reward sorted out. I'll also simplify the bits you mentioned above (removing GaussianDt, restore negative rewards, etc.)

**Comment by Guillemdb** [comment]
> Ahh, I see! What about oob=True (out-of-bounds) for boundary conditions, and terminal=True for game-ended states? OOB is something you'd see frequently in games, and Terminal is used in a bunch of the existing RL literature to mean an ended state.

I like it, thanks! I will implement it tomorrow, and I will add support for terminal conditions on the `Swarm`. Fell free to use the names that you suggested, because it less than 24h it will be already implemented.

> I'm ultimately interested in mixing the swarm with a learned model, so I agree that this may be the optimal algorithm. When you get around to that, I'm happy to help in any way that I can.

Me too. I've done a lot of work on that, but I no longer have access to the code. Once I am done implementing the features I need for the RL algorithms, I also wanna start adding learning to the Swarms.

I am planning to code my agents in Pytorch, but I will be adding support for TF in the data generation module. If you are using TF please tell me and I will make that a priority.

**Comment by justindujardin** [comment]
Thanks again for your guidance and help to integrate mathy with fragile. Fixing the oobs and ends helped a great deal! Perhaps unsurprisingly, the swarm can take on almost all of the Mathy environments and difficulties without failing! I think at this point my demo is done enough that I can clean it up and post something about it. I'll do that by this weekend.

The next step for me would be to integrate fragile as the default solver for mathy when you don't have a model installed, but I'll have to wait to do that part until you're ready to switch the license to something more compatible.

Even as we wait for the right licensing, I've created a demo that uses fragile + mathy and sent you a google colaboratory invite. You may want to check your spam because it came from one of my older email accounts. 

> The can also be visualized in a bokeh server (totally a PITA though). I haven't thought about this until you mentioned it 😅 . Plotting interactive visualizations with data streaing outside the notebook is impossible, but I could write a wrapper saves the plots as .png files in a target directory. Would that work for you?

Visualizations outside of the notebook aren't very important for me right now, but I think if you wrapped the bokeh server and launched a web browser to it while running the visualizers, that could be a good experience.

> I've done a lot of work on that, but I no longer have access to the code. Once I am done implementing the features I need for the RL algorithms, I also wanna start adding learning to the Swarms.

I wonder if there's a good way for me to try that kind of stuff out in mathy? It already has an A3C agent that is trainable. I wonder if you could describe the rough approach you would like to use to integrate the swarm with an actor-critic setup? I was thinking that perhaps the swarm could be used at each environment step to "feel out" up to n timesteps in the future and select the most appropriate action from there? But I think that could also be a really naive way of doing things.

> I am planning to code my agents in Pytorch, but I will be adding support for TF in the data generation module. 

How will the data generation module work? Do you just harvest a bunch of the high-value end-states from the HistoryTree and backtrack their paths to build episodes?

> If you are using TF please tell me and I will make that a priority.

Mathy does use TensorFlow 2.x and Keras for its agents, but I would like to move away to PyTorch because the experience has not been great with TF.

Anyway, I'm really amazed by your work! 👏 🙇 I'll continue to study how the swarm works, and think about ways to further use it in Mathy.

**Comment by Guillemdb** [comment]
> Thanks again for your guidance and help to integrate mathy with fragile. Fixing the oobs and ends helped a great deal! Perhaps unsurprisingly, the swarm can take on almost all of the Mathy environments and difficulties without failing! I think at this point my demo is done enough that I can clean it up and post something about it. I'll do that by this weekend.

I'm glad it's working. I hope I can provide better algorithms soon :)

> Thanks again for your guidance and help to integrate mathy with fragile. Fixing the oobs and ends helped a great deal! Perhaps unsurprisingly, the swarm can take on almost all of the Mathy environments and difficulties without failing! I think at this point my demo is done enough that I can clean it up and post something about it. I'll do that by this weekend.

I will be updating the license to MIT tomorrow. 

> Visualizations outside of the notebook aren't very important for me right now, but I think if you wrapped the bokeh server and launched a web browser to it while running the visualizers, that could be a good experience.

I am writing a wrapper that exports the plots as images/HTML to the provided path. I hope I can upload it tomorrow along with support for the matplotlib and plotly backends. I will also be writing tutorials for that soon.

> I was thinking that perhaps the swarm could be used at each environment step to "feel out" up to n timesteps in the future and select the most appropriate action from there?

This is the first new `Swarm` I will be uploading after I finish the couple PRs I got pending to push. It should be ready this week.

> How will the data generation module work? Do you just harvest a bunch of the high-value end-states from the HistoryTree and backtrack their paths to build episodes?

This is the first stuff that I will be publishing for the data generation module. A new tree that allows to save the data that you want for training and that makes it accessible in the way you need . I was planning to build several iterators that provide the data as TF / pytorch batched tensors ready to feed to the nets for training. For example the data could be provided as:

- (obs, action, rewards, ends)
- (obs, action, rewards, new_obs, end)
- Other formats that may be of interest
- Returning either rollouts in order or randomly sampled states from played games.
- Any other feature that you may find useful.

> It already has an A3C agent that is trainable. I wonder if you could describe the rough approach you would like to use to integrate the swarm with an actor-critic setup? 

When the basic features of the data generation feature are ready I will take a look at your Mathy agents to see what it's the easier way to pass the sampled data.

I'm still thinking the right way to do it in the general case. I want to add the option for DNN policies to be used to sample actions both in on-policy and off-policy algorithms. Doing that while allowing for running the algorithms also in a distributed fashion is not trivial, so I will need some time to figure out this. I will code the basic data generation features while I figure this out.

**Comment by Guillemdb** [comment]
I found a few bugs and I got caught up fixing them. I didn't have time to change the license. If you need it already feel free to make a PR and I will merge it.

Otherwise I will do it tomorrow.

**Comment by justindujardin** [comment]
I'm excited to hear that you're ready to make the switch to MIT license! 👏  I think integrating fragile as the default solver in Mathy will be a big win. I'll definitely do that this weekend and give you/fragile props online! 🎸 

I'm not in a hurry for the license update, but I put together a PR for you in case you don't want to deal with it: https://github.com/FragileTech/fragile/pull/73. If you want to make the changes, feel free to close the PR

**Comment by Guillemdb** [comment]
Thank you very much for the PR. It's already merged.

 I am really happy that finally someone is using my work (at least you are the first to say it in public) and contributing to the project 😁

I only need one more feature before adding a new algorithm, I think it can be ready tomorrow. I will write you when it's uploaded and link the PR in this issue.

**Comment by Guillemdb** [comment]
@justindujardin I uploaded a new kind of swarm that builds a tree for every action. Maybe it can improve the results you are getting. To use it just change `Swarm` with `StepSwarm`. 

I am not sure which strategy will work better but try passing both `FollowBestModel` and `StepToBest` model as `root_model` parameter. If your environment runs for few steps set the `step_epochs` parameter big enough to reach a terminal state in all the searches you run.

Docs: https://fragile.readthedocs.io/en/latest/module_docs/algorithms.html
Example: https://github.com/FragileTech/fragile/blob/master/tests/algorithms/test_step_swarm.py (I have not published a proper example yet)

**Comment by justindujardin** [comment]
Hey @Guillemdb, sorry for the long delay in getting back to you. Things have been busy in my life recently 😓

I managed to get fragile integrated with the latest version of Mathy (see: https://github.com/justindujardin/mathy/pull/33) so now you can run it directly from pypi:

```bash
pip install mathy[fragile]
mathy simplify "12y + 3x + 7 + 10y + 2x" --swarm
007 | -- cs -- -- ag -- | 64 | -01 | 0.0   | initial-state             | 12y + 3x + 7 + 10y + 2x
007 | -- cs -- -- ag -- | 63 | 007 | -0.01 | associative group         | 12y + 3x + (7 + 10y) + 2x
008 | -- cs -- df ag -- | 62 | 013 | -0.01 | commutative swap          | 12y + 3x + 2x + (7 + 10y)
009 | ca cs dm -- ag -- | 61 | 007 | 0.01  | distributive factoring    | 12y + (3 + 2) * x + (7 + 10y)
005 | -- cs -- -- ag -- | 60 | 005 | 0.01  | constant arithmetic       | 12y + 5x + (7 + 10y)
005 | -- cs -- -- ag -- | 59 | 003 | -0.01 | associative group         | 12y + (5x + (7 + 10y))
006 | -- cs -- df ag -- | 58 | 003 | -0.01 | commutative swap          | 5x + (7 + 10y) + 12y
007 | ca cs dm -- ag -- | 57 | 009 | 0.01  | distributive factoring    | 5x + 7 + (10 + 12) * y
003 | -- cs -- -- ag -- | 56 | 007 | 1.2   | constant arithmetic       | 5x + 7 + 22y
```
I'm going to post a tweet about it later, but I wanted to fix something first: I noticed that the pillow-simd dependency doesn't install properly on OSX unless you have extra things installed. It looks like that dependency isn't strictly needed for what Mathy does, would you be okay if I submitted a PR to remove it from the core package and put it in "atari" or another extras?

I tried out the StepSwarm briefly but it had some problems. I'll run it again with the latest version and report back specifically what's broken.

I see you're working on neural network stuff recently! 💯 I look forward to having some time to review it and see how it could fit in with Mathy

**Comment by Guillemdb** [comment]
> Things have been busy in my life recently

Hi! nice to hear from you again, I hope everything is OK :)
> I managed to get fragile integrated with the latest version of Mathy (see: justindujardin/mathy#33) so now you can run it directly from pypi:

Cool! I'm very happy to hear that!

> I noticed that the pillow-simd dependency doesn't install properly on OSX unless you have extra things installed. It looks like that dependency isn't strictly needed for what Mathy does, would you be okay if I submitted a PR to remove it from the core package and put it in "atari" or another extras?

Sure it is only used for resizing image frames, so it can be moved to Atari with no side efects.

> I tried out the StepSwarm briefly but it had some problems. I'll run it again with the latest version and report back specifically what's broken.

This is probably my fault. I have been making a lot of changes lately to simplify the overall API and maybe I broke something that is not covered by the tests. I will look into it this weekend. From now on the whole library should be way more stable, specially the core module. 

I improved a lot the `HistoryTree` to make it easier to produce examples that can be fed directly to a model, and now I am almost done with a distributed example generator. I also changed a bit the `Environment` interface to make it easier to add new environments, and allow for out-of the box parallelization. Once I am done with the features I am about to push, I will make a PR to `mathy` to improve the environment you have.

> I see you're working on neural network stuff recently! 100 I look forward to having some time to review it and see how it could fit in with Mathy

I am currently helping a friend implement a [DQDF](https://github.com/Zeta36/FractalExplorationImitationLearning) agent (DQN for supervised examples) that uses fragile to generate trajectories. 
It is still in a very preliminary phase, but once it's done I will use it as a baseline to define an agent training API.
